SonarScanner.Shim/PropertiesFileGenerator.cs

```diff
@@ -102,48 +102,125 @@ public bool TryWriteProperties(PropertiesWriter writer, out IEnumerable<ProjectD
-                .Select(g => ToProjectData(g, rootProjectBaseDir))
+                .Select(ToProjectData)
                 .ToList();
 
             var validProjects = allProjects
                 .Where(d => d.Status == ProjectInfoValidity.Valid)
                 .ToList();
 
-            if (validProjects.Any())
+            if (validProjects.Count == 0)
+            {
+                logger.LogError(Resources.ERR_NoValidProjectInfoFiles);
+                return false;
+            }
+
+            var rootProjectBaseDir = ComputeRootProjectBaseDir(projectPaths);
+            var rootModuleFiles = PutFilesToRightModuleOrRoot(validProjects, rootProjectBaseDir).ToList();
+            PostProcessProjectStatus(validProjects);
+
+            if (rootModuleFiles.Count == 0 && validProjects.All(p => p.Status == ProjectInfoValidity.NoFilesToAnalyze))
             {
-                var sharedFilePaths = validProjects
-                    .SelectMany(p => p.ExternalFiles)
-                    .Where(f => !PathHelper.IsPartOfAProject(f, projectPaths))
-                    .Distinct(StringComparer.InvariantCultureIgnoreCase)
-                    .ToList();
+                logger.LogError(Resources.ERR_NoValidProjectInfoFiles);
+                return false;
+            }
 
-                writer.WriteSonarProjectInfo(rootProjectBaseDir);
-                writer.WriteSharedFiles(sharedFilePaths);
+            writer.WriteSonarProjectInfo(rootProjectBaseDir);
+            writer.WriteSharedFiles(rootModuleFiles);
+            validProjects.ForEach(p => writer.WriteSettingsForProject(p, p.CoverageAnalysisExists(logger)
+                ? p.VisualStudioCoverageLocation : null));
 
-                foreach (var projectData in validProjects)
+            // Handle global settings
+            writer.WriteGlobalSettings(analysisProperties);
+
+            return true;
+        }
+
+        private IEnumerable<string> PutFilesToRightModuleOrRoot(IEnumerable<ProjectData> projects, string rootProjectBaseDir)
+        {
+            var fileWithProjects = projects
+                .SelectMany(p => p.ReferencedFiles.Select(f => new { Project = p, File = f }))
+                .GroupBy(group => group.File, StringComparer.OrdinalIgnoreCase)
+                .ToDictionary(group => group.Key, group => group.Select(x => x.Project).ToList());
+
+            foreach (var group in fileWithProjects)
+            {
+                var file = group.Key;
+
+                if (!File.Exists(file))
                 {
-                    writer.WriteSettingsForProject(projectData,
-                        projectData.CoverageAnalysisExists(logger) ? projectData.VisualStudioCoverageLocation : null);
+                    logger.LogWarning(Resources.WARN_FileDoesNotExist, file);
+                    logger.LogDebug(Resources.DEBUG_FileReferencedByProjects, string.Join("', '",
+                        group.Value.Select(x => x.Project.GetProjectDirectory())));
```

Comment:
_GetProjectDirectory_ is just the directory, isn't it? It would be better if the project file name was included to i.e. _Project.FullPath_.